# Comparison between the BF and Modified BF Heuristics

# Group Members

### 1- Faisal Alzahrani
### 2- Rashed Abudujen
### 3- Mohammed Alwehaibi

# Introduction

In this report, we are going to compare the backward-forward algorithm with the modified backward-forward algorithm to see which one has a better quality, and less computational effort. Also, we will test the two algorithms with given inputs to see how they compare to the optimal solution

# BF Heuristic 

## Backward Phase

In [208]:
import time # To calculate the computational time of each algorithm
start_time = time.time() 

n = 5 # Number of jobs
P = {1:24,2:13,3:32,4:9,5:3} # Processing times
D  = {1:38,2:22,3:61,4:47,5:8} # Due dates
E = {1:0,2:2,3:1,4:2,5:2} # Early penalties
L = {1:2,2:6,3:6,4:2,5:2} # Late penalties
T = sum(P.values()) # Sum of processing times for unscheduled jobs

Penalty = {1:None,2:None,3:None,4:None,5:None}
B_Sequence = []

In [209]:
num_iterations_backward = 0

for i in range(n):

 # We update the total processing times for unscheduled jobs in each iteration

    T = sum(P.values())


    for i in range(n):
        if T > D[i+1]:
            
            # Since in each iteration, we delete an element in Penalty, we have to make sure that (i+1) in Penalty
            
            if (i+1) in Penalty: # The late case
                Penalty[i+1] = ((T-D[i+1])*L[i+1])
        else:
            if (i+1) in Penalty:# The early case
                Penalty[i+1] = ((D[i+1]-T)*E[i+1])
                
    # Candidates invloves the jobs having the maximum penalty. If no tie exits, it will have only one element
    # It's useful for the case when we have a tie

    Candidates = {1:0,2:0,3:0,4:0,5:0}
    for i in range(n):

        if len(B_Sequence) == n: # If all the jobs are scheduled, we stop running
            break
        elif Penalty.get(i+1) == min(Penalty.values()):
            
            Candidates[i+1]= P[i+1]
            
        num_iterations_backward +=1


    # Append the number of the job associated with the larget processing time
    
    Jop_with_largest_P = list(Candidates.keys())[list(Candidates.values()).index(max(Candidates.values()))]
    B_Sequence.append(Jop_with_largest_P)
    P.pop(Jop_with_largest_P)
    Penalty.pop(Jop_with_largest_P)
    
    
B_Sequence.reverse() # Unlike the algorithm, we were appending from left to right. So, we need to reverse

In [210]:
num_iterations_backward

25

In [211]:
B_Sequence

[5, 2, 1, 3, 4]

## Forward Phase

In [212]:
def Calc_Pen(x):
    
    C = {0:0,1:0,2:0,3:0,4:0,5:0}
    P = {1:24,2:13,3:32,4:9,5:3}

    Penalty_f = []
    for i in range(n):

        C[i+1] = P.get(x[i]) + C[i]
        if C[i+1] > D.get(x[i]):

            Penalty_f.append(((C[i+1]-D.get(x[i]))*L.get(x[i]))) 
        else:
            Penalty_f.append(((D.get(x[i])-C[i+1])*E.get(x[i])))
   
    total_pen = sum(Penalty_f)            
    return total_pen  

In [213]:
Calc_Pen(B_Sequence)

160

In [214]:
def swap(x,i,k):
    x[i], x[i+k] = x[i+k], x[i]
    return x


In [215]:
k = n-1


In [216]:
Total_Penalty = Calc_Pen(B_Sequence)
num_iterations_forward = 0
# k represents the lag, j represents the position in the sequence

j = 0

while (j+k <= n-1): # The condition is to insure that the position of the switcing job j with lag k is possible
    
    swap(B_Sequence,j,k)
    new_Penalty =  Calc_Pen(B_Sequence)
    
    # if we get a better sequence, we need to reset the lag k to n-1, and start switching from position 0
    
    if new_Penalty < Total_Penalty:
        k = n-1
        j = 0
        Total_Penalty = new_Penalty
    
    else:
       
        # if we don't get a better sequence, we switch back to the previous sequence
        
        swap(B_Sequence,j,k)
       
        # if we have another swich with the same lag, we increment j by 1, and repeat the process. If k=0, we stop
        
        
        #Otherwise, we decrement the lag by 1, and start switching from position 0
        
        if j+1+k > n-1:
            
            if k > 0:
                k -= 1
                j = 0
            
            else:
                break
                    
        
        else:
            j+=1
            
    num_iterations_forward += 1

In [217]:
num_iterations_forward

29

In [218]:
k

0

In [219]:
Calc_Pen(B_Sequence)

141

In [220]:
B_Sequence

[5, 2, 3, 4, 1]

In [221]:
print('Time taken = %s  seconds' %(time.time()-start_time))

Time taken = 1.8416697978973389  seconds


# Modified BF Heuristic 

## Modified Backward Phase

In [222]:
start_time = time.time() 
n = 5 # Number of jobs
P = {1:24,2:13,3:32,4:9,5:3} # Processing times
D  = {1:38,2:22,3:61,4:47,5:8} # Due dates
E = {1:0,2:2,3:1,4:2,5:2} # Early penalties
L = {1:2,2:6,3:6,4:2,5:2} # Late penalties
T = sum(P.values()) # Sum of processing times for unscheduled jobs

Penalty = {1:None,2:None,3:None,4:None,5:None}
B_Sequence = []

In [223]:
num_iterations_mod_back = 0

for i in range(n):
    
    # We update the total processing times for unscheduled jobs in each iteration
    
    T = sum(P.values())


    for i in range(n):
        if T > D[i+1]: # The late case
            
            # Since in each iteration, we delete an element in Penalty, we have to make sure that (i+1) in Penalty
            
            if (i+1) in Penalty:
                
                # For the late penalty, we will have negative results, which makes it easier the apply the algorithm
                # just by taking the maximum penalty in each iteration
                
                Penalty[i+1] = ((D[i+1]-T)*L[i+1])
                
        else: # The early case
            if (i+1) in Penalty:
                Penalty[i+1] = ((D[i+1]-T)*E[i+1])

    # Candidates invloves the jobs having the maximum penalty. If no tie exits, it will have only one element
    # It's useful for the case when we have a tie

    Candidates = {1:0,2:0,3:0,4:0,5:0}
    for i in range(n):

        if len(B_Sequence) == n: # If all the jobs are scheduled, we stop running
            break
        elif Penalty.get(i+1) == max(Penalty.values()):
            
            Candidates[i+1]= P[i+1]
    
        num_iterations_mod_back +=1

    # Append the number of the job associated with the larget processing time
    
    Jop_with_largest_P = list(Candidates.keys())[list(Candidates.values()).index(max(Candidates.values()))]
    B_Sequence.append(Jop_with_largest_P)
    P.pop(Jop_with_largest_P)
    Penalty.pop(Jop_with_largest_P)
    
    
B_Sequence.reverse() # Unlike the algorithm, we were appending from left to right. So, we need to reverse

In [224]:
num_iterations_mod_back

25

In [225]:
B_Sequence

[5, 2, 1, 3, 4]

## Forward Phase

#### Same as in the Backward-Forward Heuristic 

In [226]:
def Calc_Pen(x):
    
    C = {0:0,1:0,2:0,3:0,4:0,5:0}
    P = {1:24,2:13,3:32,4:9,5:3}

    Penalty_f = []
    for i in range(n):

        C[i+1] = P.get(x[i]) + C[i]
        if C[i+1] > D.get(x[i]):

            Penalty_f.append(((C[i+1]-D.get(x[i]))*L.get(x[i]))) 
        else:
            Penalty_f.append(((D.get(x[i])-C[i+1])*E.get(x[i])))
   
    total_pen = sum(Penalty_f)            
    return total_pen  

In [227]:
Calc_Pen(B_Sequence)

160

In [228]:
def swap(x,i,k):
    x[i], x[i+k] = x[i+k], x[i]
    return x

In [229]:
k = n-1

In [230]:
Total_Penalty = Calc_Pen(B_Sequence)

# k represents the lag, j represents the position in the sequence

j = 0

while (j+k <= n-1): # The condition is to insure that the position of the switcing job j with lag k is possible
    
    swap(B_Sequence,j,k)
    new_Penalty =  Calc_Pen(B_Sequence)
    
    # if we get a better sequence, we need to reset the lag k to n-1, and start switching from position 0
    
    if new_Penalty < Total_Penalty:
        k = n-1
        j = 0
        Total_Penalty = new_Penalty
    
    else:
       
        # if we don't get a better sequence, we switch back to the previous sequence
        
        swap(B_Sequence,j,k)
       
        # if we have another swich with the same lag, we increment j by 1, and repeat the process. If k=0, we stop
        
        
        #Otherwise, we decrement the lag by 1, and start switching from position 0
        
        if j+1+k > n-1:
            
            if k > 0:
                k -= 1
                j = 0
            
            else:
                break
                    
        
        else:
            j+=1

In [231]:
k

0

In [232]:
Calc_Pen(B_Sequence)

141

In [233]:
B_Sequence

[5, 2, 3, 4, 1]

In [234]:
print('Time taken = %s  seconds' %(time.time()-start_time))

Time taken = 1.9430136680603027  seconds


## Global Optimal Solution

In [235]:
Penalty = Calc_Pen([1,2,3,4,5])
iterations = 0
for i in range(n):
    for j in range(n):
        if j != i:
            for k in range(n):
                if k != i and k != j:
                    for l in range(n):
                        if l != i and l!= j and l!= k:
                            for m in range(n):
                                if m != i and m!= j and m!= k and m!=l:
                                    Pen = Calc_Pen([i+1,j+1,k+1,l+1,m+1])
                                    if Pen < Penalty:
                                        Penalty = Pen
                                        Opt_Seq = [i+1,j+1,k+1,l+1,m+1]
                                    iterations += 1

In [236]:
print("Best penalty is", Penalty, "which is associated with the Sequence:", Opt_Seq, "with number of iterations = ", iterations)

Best penalty is 141 which is associated with the Sequence: [5, 2, 3, 4, 1] with number of iterations =  120


# Conclusion

In terms of the qualty of the obtained solution, we got the same final solution, which is equivalent to the optimal solution, and in terms of the computational effort, both had similar time taken with very slight difference. Also, we got the same number of iterations in both the backward and modified backward phase, which is 25 iterations, and same number of iterations in the forward phase, which is 29. So, total number of iterations to get the final soluton was 54. In order to get the global optimal solution we needed 120 iterations, which is considerably higher than the BF and Modified BF heuristics. In conclusion, the two algorithms performed the same in terms of the quality and computational effort, and both of them provided the optimal solution